streams/shared/src/main/scala/zio/stream/experimental/ZSink.scala

```diff
@@ -215,6 +216,70 @@ abstract class ZSink[-R, +E, -I, +Z] private (
       }
     }
 
+  /**
+   * A named alias for `race`.
+   */
+  final def |[R1 <: R, E1 >: E, A0, I1 <: I, Z1 >: Z](
+    that: ZSink[R1, E1, I1, Z1]
+  ): ZSink[R1, E1, I1, Z1] =
+    self.race(that)
+
+  /**
+   * Runs both sinks in parallel on the input, returning the result from the
+   * one that finishes successfully first.
+   */
+  final def race[R1 <: R, E1 >: E, A0, I1 <: I, Z1 >: Z](
+    that: ZSink[R1, E1, I1, Z1]
+  ): ZSink[R1, E1, I1, Z1] =
+    self.raceBoth(that).map(_.merge)
+
+  /**
+   * Runs both sinks in parallel on the input, returning the result from the
+   * one that finishes successfully first.
+   */
+  final def raceBoth[R1 <: R, E1 >: E, A0, I1 <: I, Z1](
+    that: ZSink[R1, E1, I1, Z1]
+  ): ZSink[R1, E1, I1, Either[Z, Z1]] = {
```

Comment:
Can't we use the `ZIO#race*` combinators in some fashion?
```scala
for {
  p1 <- self.push
  p2 <- that.push
  push = (in: Option[Chunk[I1]]) => p1(in).raceEither(p2(in))
} yield push
```
☝️ this is probably too good to be true, but just for the sake of the argument.

Reply:
Generally we need results from both sides because the first result may not yield a value.

`raceWith` in theory could in theory save from waiting for one value at the very last step...

`zipPar` has been used in the old implementation so chances to solve it with a one-liner are low.

Reply:
gotcha. yeah I'm pretty sure the new encoding is more amenable to these idioms than the previous sink state machinery. Perhaps:
```scala
p1(in).raceWith(p2(in))(
  (exit1, fiber2) => exit1.foldM(
    Cause.sequenceCauseOption(_).fold(fiber2.join)(ZIO.halt(_)),
    a => ZIO.succeedNow(Left(a))
  ),
  (exit2, fiber1) => ...
)
```
Just a thought though.

Reply:
`raceWith` will definitely save a whole lot of boilerplate on this combinator. Worth a try!

Reply:
@iravid it is already based on `zipPar`

Reply:
Yes - I saw that - it is worth trying to utilize raceWith there as the logic you've implemented is very similar.